contracts/modules/v2/StakingPool.sol

```diff
@@ -507,6 +507,131 @@ contract StakingPool is IStakingPool, ERC721 {
+  /// NFTs cannot be extended.
+  ///
+  /// @param tokenId           The id of the NFT that proves the ownership of the deposit.
+  /// @param initialTrancheId  The id of the tranche the deposit is already a part of.
+  /// @param newTrancheId      The id of the new tranche determining the new deposit period.
+  /// @param topUpAmount       An optional amount if the user wants to also increase the deposit
+  ///                          amount.
+  function extendDeposit(
+    uint tokenId,
+    uint initialTrancheId,
+    uint newTrancheId,
+    uint topUpAmount
+  ) public {
+    uint _firstActiveTrancheId = block.timestamp / TRANCHE_DURATION;
+    uint maxTranche = _firstActiveTrancheId + MAX_ACTIVE_TRANCHES;
+    require(tokenId != 0, "StakingPool: Invalid token id");
+    require(
+      _isApprovedOrOwner(msg.sender, tokenId),
+      "StakingPool: Not authorized to extend deposits on this token"
+    );
+    require(
+      initialTrancheId < newTrancheId,
+      "StakingPool: The chosen tranche cannot end before the initial one"
+    );
+    require(newTrancheId <= maxTranche, "StakingPool: The chosen tranche is not available yet");
+    require(
+      newTrancheId >= _firstActiveTrancheId,
+      "StakingPool: The chosen tranche has already reached the maturity date"
+    );
+
+    // If the intial tranche is expired, withdraw everything and make a new deposit equal to the
+    // withdrawn stake amount plus a top up amount if applicable. This will require the user to
+    // grant sufficient allowance beforehand.
+    if (initialTrancheId < _firstActiveTrancheId) {
+      uint[] memory trancheIds = new uint[](1);
+      trancheIds[0] = initialTrancheId;
+
+      WithdrawParams[] memory withdrawParams = new WithdrawParams[](1);
+      withdrawParams[0] = WithdrawParams(
+        tokenId,
+        true, // Withdraw deposit
+        true, // Withdraw rewards
+        trancheIds
+      );
+
+      (uint withdrawnStake, /* uint rewardsToWithdraw */) = withdraw(withdrawParams);
+
+      depositTo(
+        withdrawnStake + topUpAmount,
+        newTrancheId,
+        tokenId,
+        msg.sender
+      );
+
+      return; // Done! Skip the rest of the function.
+    }
+
+    // If the initial tranche is still active, move all the shares and pending rewards to the
+    // newly chosen tranche and its coresponding deopsit.
+
+    // First make sure tranches are up to date in terms of accumulated NXM rewards.
+    updateTranches();
+
+    Tranche memory initialTranche = tranches[initialTrancheId];
+    Tranche memory newTranche = tranches[newTrancheId];
+    Deposit memory initialDeposit = deposits[tokenId][initialTrancheId];
+    Deposit memory newDeposit = deposits[tokenId][newTrancheId];
+    uint _accNxmPerRewardsShare = accNxmPerRewardsShare;
+
+
+    // [todo] Reward shares should account for inflating the shares, using
+    // calculateRewardSharesAmount and based on the difference between the
```

Comment:
should we merge the other PR in nexus-v2 now and then merge into this one to finalize?